# Matrix Inversion

The inverse of a matrix $A$ is a matrix $A^{-1}$ such that $A \cdot A^{-1} = A^{-1} \cdot A = I$, where $I$ is the identity matrix.

**Example 1:**

In [ ]:
import sympy as sp # import sympy library
k1 = sp.Matrix([[2, 1, 2], [5, 3, 1], [1, 1, 5]]) # create matrix k1
k1

Matrix([
[2, 1, 2],
[5, 3, 1],
[1, 1, 5]])

In [ ]:
inverse_k1 = k1.inv() # calculate the inverse of matrix k1
inverse_k1

Matrix([
[7/4, -3/8, -5/8],
[ -3,    1,    1],
[1/4, -1/8,  1/8]])

In [ ]:
k1 * inverse_k1 # verify if multiplying the matrix by its inverse yields the identity matrix

Matrix([
[1, 0, 0],
[0, 1, 0],
[0, 0, 1]])

**Example 2:**

In [ ]:
import sympy as sp
a, b, c, d, e, f, g, h, i = sp.symbols('a b c d e f g h i') # define variables a-i
k = sp.Matrix([[a, b, c], [d, e, f], [g, h, i]]) # create matrix k
inverse_k = k.inv()
inverse_k

Matrix([
[ (e*i - f*h)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g), (-b*i + c*h)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g),  (b*f - c*e)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g)],
[(-d*i + f*g)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g),  (a*i - c*g)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g), (-a*f + c*d)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g)],
[ (d*h - e*g)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g), (-a*h + b*g)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g),  (a*e - b*d)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g)]])

This complex form results from the formula that allows for the calculation of the elements of the inverse matrix. Before we proceed, let's define a few concepts.

**Matrix Transposition** is an operation that swaps rows with columns. The transpose of a matrix $A$ is a matrix $A^T$ such that $A^T_{ij} = A_{ji}$.

In [ ]:
k

Matrix([
[a, b, c],
[d, e, f],
[g, h, i]])

In [ ]:
k.transpose() # transpose matrix k

Matrix([
[a, d, g],
[b, e, h],
[c, f, i]])

**Minor** of a matrix $A$ is the determinant of the matrix obtained by removing the $i$-th row and $j$-th column. We denote it by $M_{ij}$.

In [ ]:
# Minor of matrix $A$ is the determinant of the matrix obtained by removing the $i$-th row and $j$-th column. We denote it by $M_{ij}$.
k2 = sp.Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
k2

Matrix([
[1, 2, 3],
[4, 5, 6],
[7, 8, 9]])

In [ ]:
M01 = k2.minor_submatrix(0, 1) # calculate minor M01
M01

Matrix([
[4, 6],
[7, 9]])

In [ ]:
M22 = k2.minor_submatrix(2, 2) # calculate minor M22
M22

Matrix([
[1, 2],
[4, 5]])

**Cofactor** of element $a_{ij} = (-1)^{i+j} \, M_{ij}$, which is the minor multiplied by $(-1)^{i+j}$.

**Rank of a Matrix** is the largest degree of a non-zero minor of the matrix. The rank of a matrix is equal to the number of linearly independent rows or columns in the matrix.

In [ ]:
k2.rank() # calculate the rank of matrix k2

2

### Inverse of a 3x3 Matrix

Using the transpose of the cofactor matrix (minors):

$$K^{-1} =
\begin{bmatrix}
a & b & c \\
d & e & f \\
g & h & i
\end{bmatrix}^{-1}
= \frac{1}{\det(\mathbf{K})}
\begin{bmatrix}
A & B & C \\
D & E & F \\
G & H & I
\end{bmatrix}^T
= \frac{1}{\det(\mathbf{A})}
\begin{bmatrix}
A & D & G \\
B & E & H \\
C & F & I
\end{bmatrix}
$$

If the determinant is non-zero, the matrix is invertible, and the elements of the intermediate matrix on the right-hand side of the above formula are determined by:

$$
\begin{aligned}
A &= \text{det} \left( \begin{bmatrix} e & f \\ h & i \end{bmatrix} \right) & B &= -\text{det} \left( \begin{bmatrix} d & f \\ g & i \end{bmatrix} \right) & C &= \text{det} \left( \begin{bmatrix} d & e \\ g & h \end{bmatrix} \right) \\
D &= -\text{det} \left( \begin{bmatrix} b & c \\ h & i \end{bmatrix} \right) & E &= \text{det} \left( \begin{bmatrix} a & c \\ g & i \end{bmatrix} \right) & F &= -\text{det} \left( \begin{bmatrix} a & b \\ g & h \end{bmatrix} \right) \\
G &= \text{det} \left( \begin{bmatrix} b & c \\ e & f \end{bmatrix} \right) & H &= -\text{det} \left( \begin{bmatrix} a & c \\ d & f \end{bmatrix} \right) & I &= \text{det} \left( \begin{bmatrix} a & b \\ d & e \end{bmatrix} \right)
\end{aligned}
$$

These are the determinants of 2x2 matrices formed by removing the appropriate rows and columns and taking into account the sign $(-1)^{i+j}$.

### Inverse of a 2x2 Matrix

Similarly, for a 2x2 matrix, the inverse can be calculated using the cofactor matrix:

$$
\mathbf{L}^{-1} =
\begin{bmatrix}
a & b \\
c & d
\end{bmatrix}^{-1}
= \frac{1}{\det(\mathbf{L})}
\begin{bmatrix}
D & C \\
B & A
\end{bmatrix}^T
= \frac{1}{\det(\mathbf{L})}
\begin{bmatrix}
D & B \\
C & A
\end{bmatrix}
$$

where $\det(\mathbf{L}) = ad - bc$

If the determinant is non-zero, the matrix is invertible, and the elements of the intermediate matrix are given as:

$$
\begin{aligned}
A &= \text{det} \begin{bmatrix} d \end{bmatrix} = d, & B &= -\text{det} \begin{bmatrix} b \end{bmatrix} = -b, & C &= -\text{det} \begin{bmatrix} c \end{bmatrix} = -c, &
D &= \text{det} \begin{bmatrix} a \end{bmatrix} = a.
\end{aligned}
$$

These elements are determined by calculating the determinants of 1x1 matrices formed by removing the appropriate rows and columns from the original 2x2 matrix and applying the corresponding signs $(-1)^{i+j}$.

---

## Exercises for Students

1. Find the inverse matrix for

$$
\begin{bmatrix}
2 & 0 & 1 \\
0 & 1 & 0 \\
1 & 2 & 0
\end{bmatrix}
$$

and verify if the result is correct.

2. Using the `det()` command, determine the rank of the matrix below
$$
A =
\begin{bmatrix}
4 & -3 & 7 \\
-1 & 6 & 3 \\
2 & 9 & 1
\end{bmatrix}
$$

In [ ]:
from sympy import Matrix, symbols

# Question 1: Finding the inverse matrix
matrix_1 = np.array([[2, 0, 1], [0, 1, 0], [1, 2, 0]])
inverse_matrix_1 = np.linalg.inv(matrix_1)
result_1 = np.dot(matrix_1, inverse_matrix_1)  # Verify by multiplying the original and the inverse

# Question 2: Determining the rank using det() for matrix A
matrix_A = np.array([[4, -3, 7], [-1, 6, 3], [2, 9, 1]])
det_A = np.linalg.det(matrix_A)
rank_A = np.linalg.matrix_rank(matrix_A)

inverse_matrix_1, result_1, det_A, rank_A


(array([[ 0., -2.,  1.],
        [ 0.,  1.,  0.],
        [ 1.,  4., -2.]]),
 array([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]]),
 -251.9999999999998,
 3)

### Matrix Problems

---

## Problem 1: Find the Inverse of a Matrix and Verify the Result

Given the matrix:
$$
\mathbf{M} =
\begin{bmatrix}
2 & 0 & 1 \\
0 & 1 & 0 \\
1 & 2 & 0
\end{bmatrix}
$$

### Step 1: Compute the Determinant
The determinant of \( \mathbf{M} \) is:
$$
\text{det}(\mathbf{M}) =
\begin{vmatrix}
2 & 0 & 1 \\
0 & 1 & 0 \\
1 & 2 & 0
\end{vmatrix}
= 2 \begin{vmatrix} 1 & 0 \\ 2 & 0 \end{vmatrix}
- 0 \begin{vmatrix} 0 & 0 \\ 1 & 0 \end{vmatrix}
+ 1 \begin{vmatrix} 0 & 1 \\ 1 & 2 \end{vmatrix}
$$

Compute the minors:
$$
\begin{vmatrix} 1 & 0 \\ 2 & 0 \end{vmatrix} = (1)(0) - (0)(2) = 0
$$
$$
\begin{vmatrix} 0 & 1 \\ 1 & 2 \end{vmatrix} = (0)(2) - (1)(1) = -1
$$

Substitute back:
$$
\text{det}(\mathbf{M}) = 2(0) + 0 + 1(-1) = -1
$$

Since \( \text{det}(\mathbf{M}) \neq 0 \), the matrix is invertible.

### Step 2: Compute the Adjugate Matrix
The adjugate is the transpose of the cofactor matrix. First, compute the cofactors:

1. $$C_{11} = \begin{vmatrix} 1 & 0 \\ 2 & 0 \end{vmatrix} = 0$$
2. $$ C_{12} = -\begin{vmatrix} 0 & 0 \\ 1 & 0 \end{vmatrix} = 0 $$
3. $$ C_{13} = \begin{vmatrix} 0 & 1 \\ 1 & 2 \end{vmatrix} = -1 $$
4. $$ C_{21} = -\begin{vmatrix} 0 & 1 \\ 2 & 0 \end{vmatrix} = -2$$
5. $$ C_{22} = \begin{vmatrix} 2 & 1 \\ 1 & 0 \end{vmatrix} = -1 $$
6. $$ C_{23} = -\begin{vmatrix} 2 & 0 \\ 1 & 2 \end{vmatrix} = -4 $$
7. $$ C_{31} = \begin{vmatrix} 0 & 1 \\ 1 & 0 \end{vmatrix} = -1 $$
8. $$ C_{32} = -\begin{vmatrix} 2 & 1 \\ 0 & 0 \end{vmatrix} = 0 $$
9. $$ C_{33} = \begin{vmatrix} 2 & 0 \\ 0 & 1 \end{vmatrix} = 2 $$

The cofactor matrix:
$$
\mathbf{C} =
\begin{bmatrix}
0 & 0 & -1 \\
-2 & -1 & -4 \\
-1 & 0 & 2
\end{bmatrix}
$$

The adjugate matrix is the transpose of \( \mathbf{C} \):
$$
\text{adj}(\mathbf{M}) =
\begin{bmatrix}
0 & -2 & -1 \\
0 & -1 & 0 \\
-1 & -4 & 2
\end{bmatrix}
$$

### Step 3: Compute the Inverse
The inverse of \( \mathbf{M} \) is:
$$
\mathbf{M}^{-1} = \frac{1}{\text{det}(\mathbf{M})} \cdot \text{adj}(\mathbf{M})
$$

Substitute \( \text{det}(\mathbf{M}) = -1 \):
$$
\mathbf{M}^{-1} =
\begin{bmatrix}
0 & 2 & 1 \\
0 & 1 & 0 \\
1 & 4 & -2
\end{bmatrix}
$$

### Step 4: Verify the Result
Multiply \( \mathbf{M} \) by \( \mathbf{M}^{-1} \):

$$
\mathbf{M} \cdot \mathbf{M}^{-1} =
\begin{bmatrix}
2 & 0 & 1 \\
0 & 1 & 0 \\
1 & 2 & 0
\end{bmatrix}
\cdot
\begin{bmatrix}
0 & 2 & 1 \\
0 & 1 & 0 \\
1 & 4 & -2
\end{bmatrix}
=
\begin{bmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1
\end{bmatrix}
$$

Thus, the result is verified.

---

## Problem 2: Determine the Rank of a Matrix

Given:
$$
\mathbf{A} =
\begin{bmatrix}
4 & -3 & 7 \\
-1 & 6 & 3 \\
2 & 9 & 1
\end{bmatrix}
$$

### Step 1: Compute the Determinant of \( \mathbf{A} \)
$$
\text{det}(\mathbf{A}) =
4 \begin{vmatrix} 6 & 3 \\ 9 & 1 \end{vmatrix}
- (-3) \begin{vmatrix} -1 & 3 \\ 2 & 1 \end{vmatrix}
+ 7 \begin{vmatrix} -1 & 6 \\ 2 & 9 \end{vmatrix}
$$

Calculate the minors:
$$
\begin{vmatrix} 6 & 3 \\ 9 & 1 \end{vmatrix} = (6)(1) - (3)(9) = 6 - 27 = -21
$$
$$
\begin{vmatrix} -1 & 3 \\ 2 & 1 \end{vmatrix} = (-1)(1) - (3)(2) = -1 - 6 = -7
$$
$$
\begin{vmatrix} -1 & 6 \\ 2 & 9 \end{vmatrix} = (-1)(9) - (6)(2) = -9 - 12 = -21
$$

Substitute back:
$$
\text{det}(\mathbf{A}) = 4(-21) + 3(-7) + 7(-21) = -84 - 21 - 147 = -252
$$

Since \( \text{det}(\mathbf{A}) \neq 0 \), the matrix has full rank, meaning:
$$
\text{rank}(\mathbf{A}) = 3
$$